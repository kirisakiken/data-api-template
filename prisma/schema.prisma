// This is your Prisma schema file,
// learn more about it in the docs: https://pris.ly/d/prisma-schema

generator client {
  provider = "prisma-client-js"
}

datasource db {
  provider = "postgresql"
  url      = env("DATABASE_URL")
}

enum StudentType {
  unknown
  associate
  bachelors
  masters
  doctoral
}

enum LecturerType {
  assistant_professor
  associate_professor
  full_professor
  distinguished_professor
}

model User {
  id Int @id @default(autoincrement())
  createdAt DateTime @default(now())
  updatedAt DateTime @updatedAt

  email String @unique
  hash String

  firstName String?
  lastName String?

  bookmarks Bookmark[]

  @@map("users")
}

model Bookmark {
  id Int @id @default(autoincrement())
  createdAt DateTime @default(now())
  updatedAt DateTime @updatedAt

  title String
  description String?
  link String

  user_id Int
  user User @relation(fields: [user_id], references: [id], onDelete: Cascade)

  @@map("bookmarks")
}

model Student {
  id Int @default(autoincrement())
  createdAt DateTime @default(now())
  updatedAt DateTime @updatedAt

  student_no Int @unique
  first_name String
  last_name String
  student_type StudentType
  entry_year Int

  @@id([student_no])
  @@map("students")
}

model Lecturer {
  id Int @default(autoincrement())
  createdAt DateTime @default(now())
  updatedAt DateTime @updatedAt

  lecturer_no Int @unique
  first_name String
  last_name String
  lecturer_type LecturerType
  entry_year Int
  salary Int

  lectures Lecture[]

  @@id([lecturer_no])
  @@map("lecturers")
}

model Lecture {
  id Int @default(autoincrement())
  createdAt DateTime @default(now())
  updatedAt DateTime @updatedAt

  lecture_no String @unique
  lecturer_reference Int?

  lecturer Lecturer? @relation(fields: [lecturer_reference], references: [lecturer_no])

  @@id([lecture_no])
  @@map("lectures")
}
