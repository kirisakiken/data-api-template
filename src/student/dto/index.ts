export * from './student-query-options'