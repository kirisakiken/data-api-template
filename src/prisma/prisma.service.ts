import { Injectable, NotFoundException } from '@nestjs/common'
import { ConfigService } from '@nestjs/config'
import { PrismaClient } from '@prisma/client'
import { CreateBookMarkDto, UpdateBookmarkDto } from 'src/bookmark/dto'

@Injectable()
export class PrismaService extends PrismaClient {
  constructor(private configService: ConfigService) {
    super({
      datasources: {
        db: {
          url: configService.get('DATABASE_URL'),
        },
      },
    })
  }

  clearDb() {
    return this.$transaction([
      this.bookmark.deleteMany(),
      this.user.deleteMany(),
    ])
  }

  //#region Bookmarks Repository Calls

  getAllBookmarks(userId: number) {
    return this.bookmark.findMany({
      where: {
        user_id: userId,
      },
    })
  }

  async getBookmark(userId: number, bookmarkId: number) {
    const bookmark = await this.bookmark.findFirst({
      where: {
        id: bookmarkId,
        user_id: userId,
      },
    })

    if (!bookmark) {
      throw new NotFoundException(
        `Unable to find bookmark by id: ${bookmarkId}`,
      )
    }

    return bookmark
  }

  createBookmark(userId: number, dto: CreateBookMarkDto) {
    return this.bookmark.create({
      data: {
        user_id: userId,
        ...dto,
      },
    })
  }

  async updateBookmark(
    userId: number,
    bookmarkId: number,
    dto: UpdateBookmarkDto,
  ) {
    const bookmark = await this.bookmark.findFirst({
      where: {
        id: bookmarkId,
        user_id: userId,
      },
    })

    if (!bookmark) {
      throw new NotFoundException(
        `Unable to find bookmark by id: ${bookmarkId}`,
      )
    }

    return this.bookmark.update({
      where: {
        id: bookmarkId,
      },
      data: {
        ...dto,
      },
    })
  }

  async deleteBookmarkById(userId: number, bookmarkId: number) {
    const bookmark = await this.getBookmark(userId, bookmarkId)

    if (!bookmark) {
      throw new NotFoundException(
        `Unable to find bookmark by id: ${bookmarkId}`,
      )
    }

    return this.bookmark.delete({
      where: {
        id: bookmark.id,
      },
    })
  }

  //#endregion
}
