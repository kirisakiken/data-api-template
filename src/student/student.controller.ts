import { Controller, Get, Inject, Param, ParseEnumPipe, ParseIntPipe, UseGuards } from '@nestjs/common'
import { StudentType } from '@prisma/client'
import { JwtGuard } from '../auth/guard'
import { StudentService } from './student.service'

@UseGuards(JwtGuard)
@Controller('student')
export class StudentController {
  @Inject()
  private studentService!: StudentService

  @Get()
  listStudents() {
    return this.studentService.listStudents()
  }

  @Get('/by-student-no/:student_no')
  getStudentByNo(@Param('student_no', ParseIntPipe) studentNo: number) {
    return this.studentService.getStudentByNo(studentNo)
  }

  @Get('/by-student-type/:student_type')
  getStudentsByType(@Param('student_type') studentType: StudentType) {
    return this.studentService.getStudentsByType(studentType)
  }

  @Get('/by-entry-year/:entry_year')
  getStudentsByEntryYear(@Param('entry_year', ParseIntPipe) entryYear: number) {
    return this.studentService.getStudentsByEntryYear(entryYear)
  }
}
