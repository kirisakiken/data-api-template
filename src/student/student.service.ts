import { Inject, Injectable, NotFoundException } from '@nestjs/common'
import { StudentType } from '@prisma/client'
import { PrismaService } from '../prisma/prisma.service'
import { studentSelectOptions } from './dto'

@Injectable()
export class StudentService {
  @Inject()
  private prismaService!: PrismaService

  public listStudents() {
    return this.prismaService.student.findMany({
      select: studentSelectOptions,
    })
  }

  public async getStudentByNo(studentNo: number) {
    const student = await this.prismaService.student.findUnique({
      where: {
        student_no: studentNo,
      },
      select: studentSelectOptions,
    })

    if (!student) {
      throw new NotFoundException(`Unable to find student by no: ${studentNo}`)
    }

    return student
  }

  public getStudentsByType(studentType: StudentType) {
    return this.prismaService.student.findMany({
      where: {
        student_type: studentType,
      },
      select: studentSelectOptions,
    })
  }

  public getStudentsByEntryYear(entryYear: number) {
    return this.prismaService.student.findMany({
      where: {
        entry_year: entryYear,
      },
      select: studentSelectOptions,
    })
  }
}
