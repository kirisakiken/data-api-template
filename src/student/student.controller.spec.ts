import { ConfigModule } from '@nestjs/config'
import { Test, TestingModule } from '@nestjs/testing'
import { PrismaService } from '../prisma/prisma.service'
import { StudentController } from './student.controller'
import { StudentService } from './student.service'

describe('StudentController', () => {
  // let configService: ConfigService
  let prismaService: PrismaService
  let studentController: StudentController
  let studentService: StudentService

  beforeAll(async () => {
    const module: TestingModule = await Test.createTestingModule({
      imports: [
        ConfigModule.forRoot({
          isGlobal: true,
        }),
      ],
      controllers: [StudentController],
      providers: [
        PrismaService,
        StudentService,
      ],
    }).compile()

    prismaService = module.get<PrismaService>(PrismaService)

    studentController = module.get<StudentController>(StudentController)
    studentService = module.get<StudentService>(StudentService)
  })

  beforeEach(async () => {
    await prismaService.student.createMany({
      data: [
        {
          student_no: 1,
          first_name: 's-fn-1',
          last_name: 's-ln-1',
          student_type: 'associate',
          entry_year: 2022,
        },
        {
          student_no: 2,
          first_name: 's-fn-2',
          last_name: 's-ln-2',
          student_type: 'bachelors',
          entry_year: 2022,
        },
      ]
    })
  })

  afterEach(async () => {
    await prismaService.student.deleteMany()
  })

  describe('dependencies', () => {
    it('dependencies should be defined', () => {
      expect(prismaService).toBeDefined()
  
      expect(studentController).toBeDefined()
      expect(studentService).toBeDefined()
    })
  })
  
  describe('StudentController Get', () => {
    describe('listStudents', () => {
      it('list successfully', async () => {
        const result = await studentController.listStudents()
        expect(result).toMatchSnapshot()
      })
    })

    describe('getStudentByNo', () => {
      it('get sucessfully', async () => {

      })
    })
  })
})
