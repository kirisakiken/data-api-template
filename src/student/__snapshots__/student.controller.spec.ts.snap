// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StudentController StudentController Get listStudents list successfully 1`] = `
Array [
  Object {
    "createdAt": 2022-06-08T19:51:35.260Z,
    "entry_year": 2022,
    "first_name": "s-fn-1",
    "id": 3,
    "last_name": "s-ln-1",
    "student_no": 1,
    "student_type": "associate",
    "updatedAt": 2022-06-08T19:51:35.260Z,
  },
  Object {
    "createdAt": 2022-06-08T19:51:35.260Z,
    "entry_year": 2022,
    "first_name": "s-fn-2",
    "id": 4,
    "last_name": "s-ln-2",
    "student_no": 2,
    "student_type": "bachelors",
    "updatedAt": 2022-06-08T19:51:35.260Z,
  },
]
`;
